storage/netstore.go

```diff
@@ -18,318 +18,293 @@ package storage
-	n.mu.Lock()
-	defer n.mu.Unlock()
+	n.putMu.Lock()
+	defer n.putMu.Unlock()
+
+	log.Trace("netstore.put", "ref", ch.Address().String(), "mode", mode)
 
-	// put to the chunk to the store, there should be no error
+	// put the chunk to the localstore, there should be no error
 	exists, err := n.Store.Put(ctx, mode, ch)
 	if err != nil {
 		return exists, err
 	}
 
-	// if chunk is now put in the store, check if there was an active fetcher and call deliver on it
-	// (this delivers the chunk to requestors via the fetcher)
-	log.Trace("n.getFetcher", "ref", ch.Address())
-	if f := n.getFetcher(ch.Address()); f != nil {
-		log.Trace("n.getFetcher deliver", "ref", ch.Address())
-		f.deliver(ctx, ch)
-	}
-	return exists, nil
-}
+	// notify RemoteGet (or SwarmSyncerClient) about a chunk delivery and it being stored
+	fi, ok := n.fetchers.Get(ch.Address().String())
+	if ok {
+		// we need SafeClose, because it is possible for a chunk to both be
+		// delivered through syncing and through a retrieve request
+		fii := fi.(*Fetcher)
+		fii.SafeClose()
+		log.Trace("netstore.put chunk delivered and stored", "ref", ch.Address().String())
 
-// Get retrieves the chunk from the NetStore DPA synchronously.
-// It calls NetStore.get, and if the chunk is not in local Storage
-// it calls fetch with the request, which blocks until the chunk
-// arrived or context is done
-func (n *NetStore) Get(rctx context.Context, mode chunk.ModeGet, ref Address) (Chunk, error) {
-	chunk, fetch, err := n.get(rctx, mode, ref)
-	if err != nil {
-		return nil, err
-	}
-	if chunk != nil {
-		// this is not measuring how long it takes to get the chunk for the localstore, but
-		// rather just adding a span for clarity when inspecting traces in Jaeger, in order
-		// to make it easier to reason which is the node that actually delivered a chunk.
-		_, sp := spancontext.StartSpan(
-			rctx,
-			"localstore.get")
-		defer sp.Finish()
-
-		return chunk, nil
-	}
-	return fetch(rctx)
-}
+		metrics.GetOrRegisterResettingTimer(fmt.Sprintf("netstore.fetcher.lifetime.%s", fii.CreatedBy), nil).UpdateSince(fii.CreatedAt)
 
-// FetchFunc returns nil if the store contains the given address. Otherwise it returns a wait function,
-// which returns after the chunk is available or the context is done
-func (n *NetStore) FetchFunc(ctx context.Context, ref Address) func(context.Context) error {
-	chunk, fetch, _ := n.get(ctx, chunk.ModeGetRequest, ref)
-	if chunk != nil {
-		return nil
-	}
-	return func(ctx context.Context) error {
-		_, err := fetch(ctx)
-		return err
+		// helper snippet to log if a chunk took way to long to be delivered
+		if time.Since(fii.CreatedAt) > 5*time.Second {
+			log.Trace("netstore.put slow chunk delivery", "ref", ch.Address().String())
+		}
+
+		n.fetchers.Remove(ch.Address().String())
 	}
+
+	return exists, nil
 }
 
 // Close chunk store
-func (n *NetStore) Close() (err error) {
-	close(n.closeC)
-
-	wg := sync.WaitGroup{}
-	for _, key := range n.fetchers.Keys() {
```

Comment:
what about all of this open fetches cancellation on shutdown that was removed? is this no longer needed?

Reply:
There is a comment by @zelig above, where I explain in detail about that.

Reply:
https://github.com/ethersphere/swarm/pull/1344#discussion_r276473055

the `fetchers` collection right now is just about state - we cannot cancel them from here.

we could wait as @zelig suggest, but then we introduce yet another timeout, and i don't think we need that.

maybe it makes sense to have the `fetchers` not just as a collection here, but actually manage their full lifecycle (ala the actual code that I am removing right now), but this is a major change, so I'd rather we work on it iteratively as a next step. for example `fetchers` lifecycle is started at retrieve requests and syncing... there is no way you can cancel any of these remote requests, so i am not sure what we are going to gain with adding `cancelation` here.